### ISSUE TITLE
New lint: Unix permissions set to a non-octal value

### ISSUE THREAD

**Issue by xTibor** [rust-lang/rust-clippy#6934]
### What it does

Checks if the following functions expecting Unix permission bits only called with octal literals:

* [`PermissionsExt::set_mode`](https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html#tymethod.set_mode)
* [`PermissionsExt::from_mode`](https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html#tymethod.from_mode)
* [`OpenOptionsExt::mode`](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode)
* [`DirBuilderExt::mode`](https://doc.rust-lang.org/std/os/unix/fs/trait.DirBuilderExt.html#tymethod.mode)

### Categories

- Kind: `clippy::correctness`

### Drawbacks

False positives maybe.

### Example

```rust
let mut options = OpenOptions::new();
options.mode(440);
```

This example tries to set the file permissions to `440` (`r,r,-`) however it ends up as `0o670` (`rw,rwx,-`) due to the decimal/octal confusion, giving unintended write/execute permissions in the process.

Clippy should suggest rewriting that line as `options.mode(0o440);`.

**Comment by ebobrow** [comment]
@giraffate Can I get a little help on this one? I need to be able to check if the method `mode`, for example, is being called on an object of type `OpenOptions`. Currently, my lint would apply to any object calling any method with the name `mode`. Sorry, I wasn't sure who to tag here.

**Comment by giraffate** [comment]
@ebobrow Thanks for working on this!

I think these codes may help you to detect a receiver type of `OpenOptions`:
<https://github.com/rust-lang/rust-clippy/blob/0743e841f07b7c41f65187fbf71f90bb5dc71982/clippy_lints/src/open_options.rs#L32-L43>

**Comment by ebobrow** [comment]
Thank you so much!
